File: tutor-service/app/contracts.py
from __future__ import annotations

from typing import Any

from .catalog import LESSON_CATALOG

VALID_ACTIONS = {"highlight_criterion", "clear_highlight", "pulse_tutor"}
VALID_STATUSES = {"idle", "dirty", "evaluating", "success", "failure"}


class TutorRenderCommand:
    def __init__(self, action: str, criterion_id: str | None = None) -> None:
        if action not in VALID_ACTIONS:
            raise ValueError("unsupported tutor render action")
        if action == "highlight_criterion" and not criterion_id:
            raise ValueError("highlight_criterion requires criterion_id")
        if action != "highlight_criterion" and criterion_id is not None:
            raise ValueError("only highlight_criterion accepts criterion_id")
        self.action = action
        self.criterion_id = criterion_id

    def as_dict(self) -> dict[str, str]:
        command = {"layer": "lesson_tutor", "action": self.action}
        if self.criterion_id:
            command["criterion_id"] = self.criterion_id
        return command


def validate_context(value: Any) -> dict[str, Any]:
    """Accept learner state only; derive all lesson language and tool IDs from the allowlisted catalog."""
    if not isinstance(value, dict):
        raise ValueError("context must be an object")
    lesson_id = value.get("lesson_id")
    screen_id = value.get("screen_id")
    status = value.get("status")
    if not isinstance(lesson_id, str) or not isinstance(screen_id, str):
        raise ValueError("lesson_id and screen_id are required strings")
    if status not in VALID_STATUSES:
        raise ValueError("invalid lesson status")

    screen = LESSON_CATALOG.get(lesson_id, {}).get(screen_id)
    if not screen:
        raise ValueError("lesson screen is not tutor-enabled")
    criteria = screen["criteria"]
    criterion_ids = {item["id"] for item in criteria}
    failed = value.get("failed_criteria", [])
    if not isinstance(failed, list) or any(item not in criterion_ids for item in failed):
        raise ValueError("failed criteria do not match the lesson")
    hint_level = value.get("hint_level", 0)
    if isinstance(hint_level, float) and hint_level.is_integer():
        hint_level = int(hint_level)
    # Live tutor + Finn ladder: H0 unused / H1–H2 non-answer only (no walkthrough H3–H4).
    if not isinstance(hint_level, int) or hint_level < 0 or hint_level > 2:
        raise ValueError("invalid hint level")

    return {
        "lesson_id": lesson_id,
        "screen_id": screen_id,
        "prompt": screen["prompt"],
        "criteria": criteria,
        "status": status,
        "failed_criteria": failed,
        "hint_level": hint_level,
        "learner_text": "",
    }

File: server/tutor/contracts.ts
import { LESSON_CATALOG } from "./catalog";

export const VALID_ACTIONS = new Set(["highlight_criterion", "clear_highlight", "pulse_tutor"]);
export const VALID_STATUSES = new Set(["idle", "dirty", "evaluating", "success", "failure"]);

export type TutorRenderCommand = {
  layer: "lesson_tutor";
  action: string;
  criterion_id?: string;
};

export type ValidatedLessonContext = {
  lesson_id: string;
  screen_id: string;
  prompt: string;
  criteria: { id: string; label: string }[];
  status: string;
  failed_criteria: string[];
  hint_level: number;
  learner_text: string;
};

export function makeRenderCommand(action: string, criterionId?: string | null): TutorRenderCommand {
  if (!VALID_ACTIONS.has(action)) throw new Error("unsupported tutor render action");
  if (action === "highlight_criterion" && !criterionId) {
    throw new Error("highlight_criterion requires criterion_id");
  }
  if (action !== "highlight_criterion" && criterionId != null) {
    throw new Error("only highlight_criterion accepts criterion_id");
  }
  const command: TutorRenderCommand = { layer: "lesson_tutor", action };
  if (criterionId) command.criterion_id = criterionId;
  return command;
}

/** Accept learner state only; derive prompt/criteria from allowlisted catalog. */
export function validateContext(value: unknown): ValidatedLessonContext {
  if (!value || typeof value !== "object") throw new Error("context must be an object");
  const v = value as Record<string, unknown>;
  const lessonId = v.lesson_id;
  const screenId = v.screen_id;
  const status = v.status;
  if (typeof lessonId !== "string" || typeof screenId !== "string") {
    throw new Error("lesson_id and screen_id are required strings");
  }
  if (!VALID_STATUSES.has(String(status))) throw new Error("invalid lesson status");

  const screen = LESSON_CATALOG[lessonId]?.[screenId];
  if (!screen) throw new Error("lesson screen is not tutor-enabled");

  const criteria = screen.criteria;
  const criterionIds = new Set(criteria.map((c) => c.id));
  const failed = v.failed_criteria ?? [];
  if (!Array.isArray(failed) || failed.some((item) => typeof item !== "string" || !criterionIds.has(item))) {
    throw new Error("failed criteria do not match the lesson");
  }

  let hintLevel = v.hint_level ?? 0;
  if (typeof hintLevel === "number" && Number.isInteger(hintLevel) === false && Number.isInteger(Number(hintLevel))) {
    hintLevel = Number(hintLevel);
  }
  if (typeof hintLevel !== "number" || !Number.isInteger(hintLevel) || hintLevel < 0 || hintLevel > 2) {
    throw new Error("invalid hint level");
  }

  return {
    lesson_id: lessonId,
    screen_id: screenId,
    prompt: screen.prompt,
    criteria,
    status: String(status),
    failed_criteria: failed as string[],
    hint_level: hintLevel,
    learner_text: "",
  };
}
